### ISSUE TITLE
error when using new beta OS of hybrid router

### ISSUE THREAD

**Issue by hgehrts** [melle/l33tport#6]
Hi
 I am using a beta of the Hybrid Router OS. The release is 'Speedport Hybrid 050124.03.05.015.bin'.
It seems to me that the l33tport scripts do not work anymore. They did work ok for me before with the old OS release I had before.
Now, when I start l33tport, I get this:
root@pi3:/opt/l33tport# /usr/bin/node /opt/l33tport/l33tport.js -o json -f dsl
events.js:141
      throw er; // Unhandled 'error' event
      ^

Error: socket hang up
    at createHangUpError (_http_client.js:213:15)
    at Socket.socketOnEnd (_http_client.js:305:23)
    at emitNone (events.js:72:20)
    at Socket.emit (events.js:166:7)
    at endReadableNT (_stream_readable.js:923:12)
    at nextTickCallbackWith2Args (node.js:511:9)
    at process._tickCallback (node.js:425:17)

**Comment by Joshua1337** [comment]
Where did you got the Beta?

**Comment by hgehrts** [comment]
From a Telekom technician. I had massive issues with CRC errors so he gave me the link to the beta image to test it. The box works ok now it seems but I'd rather be able to record the metrics again. 
I have no idea how I can debug this.... I have no clue of JavaScript and Node.js

**Comment by wusamba** [comment]
I think the login was changed with the new firmware (v050124.03.05.017) - Login.json didn't give a challengev anymore.

`curl -d csrf_token=nulltoken -d showpw=0 -d challengev=null https://speedport.ip/data/login.json`
`curl: (52) Empty reply from server`

**Comment by snookerfly** [comment]
Also got the new firmware (050124.03.05.017) last week and l33tport stopped working... any ideas?

**Comment by melle** [comment]
I looked into the login procedure of the new firmware, it's a bit different now: The challenge is created on the device and embedded into the source code of the login page `/html/login/index.html?lang=en`

```
  <!-- Global script data -->
  <script type="text/javascript">
    var csrf_token = "nulltoken";
    var challenge = "14484E2A4EC8aCbcf5bf70209f18edCb4E4BDE8Ef7CDDB1854869eAfA13E8bf0";
    /* <![CDATA[ */
    var rootRelative = '../';
    var JSONSource   = '../../data/Login.json';
    /* ]]> */
  </script>

```

You can confirm this by reloading the login page and observe, that the challenge is different with every request.

The login form is sent as POST containing all variables in the body:

```
POST /data/Login.json?lang=en HTTP/1.1
Host: 192.168.7.1
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.12; rv:53.0) Gecko/20100101 Firefox/53.0
Accept: application/json, text/javascript, */*
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Content-Type: application/x-www-form-urlencoded
X-Requested-With: XMLHttpRequest
Referer: http://192.168.7.1/html/login/?lang=en
Content-Length: 189
Connection: keep-alive

password_shaddowed=17cce891847485c3d3286eda78a81931f2c47a4d4b8dac009069df951dcc4e2e&csrf_token=nulltoken&showpw=1&challengev=EaedCCF820c5C889FdD7dE6ABBCd1EAF9c4cECCf9eDC3ed6820edcCC2B7aB06C
```

The response sets the session cookie:

```
HTTP/1.1 200 OK
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
Set-Cookie: SessionID_R3=11h20200Ze2; path=/; HttpOnly;

```

The login procedure is in `/js/pages/login.js?vernum=017` and looks pretty similar to the previous login. Adopting l33tport to it should not be too hard, more like a fun evening with wireshark and javascript hacking. 

Unfortunately I'm pretty busy a.t.m. with my real life, but if anyone want's to jump into the javascript mess, I would be more than thankful for a pull request 🐳

**Comment by melle** [comment]
Ok, I made a first attempt to fix the login, see the [new-login branch](https://github.com/melle/l33tport/compare/new-login).

The good news: the `challengev` handling is still the same, by using the `challenge` value from the login page, I can create the exact same password hash.

The bad news: the router refuses to respond. I assume we need to send the right headers and/or use a keep-alive connection, because the challenge could be bound to the actual connection.

**Comment by wusamba** [comment]
Maybe you can use some information from here: https://forum.stricted.net/index.php/Thread/253-version-1-0-pre13/?postID=6495#post6495

I am not firm in nodejs, but maybe it is possible to use cookies to prevent login attempts? So the data can be collected and transferred to rrdtool inside the nodejs script. Or can cookies saved and reloaded in nodejs?

Thanks for your work.

**Comment by basisbit** [comment]
@melle I can confirm that with current firmware, the master branch does not work any more. Also, the new-login branch seems to still have some problems.
example:

> nodejs ./l33tport.js -f speed
> 
> /home/share/l33tport/l33tport.js:105
>             onResult(res.statusCode, output);
>             ^
> TypeError: undefined is not a function
>     at IncomingMessage.<anonymous> (/home/share/l33tport/l33tport.js:105:13)
>     at IncomingMessage.emit (events.js:117:20)
>     at _stream_readable.js:929:16
>     at process._tickCallback (node.js:419:13)

Anything else I could provide to help this project a bit?

**Comment by foto-andreas** [comment]
@basisbit  remove the parameter "options" in the surrounding function (line 89)

but even then my sph does not respond to the password...

**Comment by dYna77** [comment]
Any info if this will work again?

**Comment by hensur** [comment]
I fixed it in #8

**Comment by melle** [comment]
Merged #8, thanks to @hensur for the pr.
